app/workers/zync_worker.rb

```diff
@@ -233,16 +234,44 @@ def valid?(event_id)
     true
   end
 
-  def publish_dependencies_events(event_id)
+  def sync_dependencies(event_id)
+    dependency_events = create_dependency_events(event_id)
+
+    publish_dependency_events(dependency_events)
+    enqueue_dependency_batch(event_id, dependency_events)
+  end
+
+  def create_dependency_events(event_id)
     event = EventStore::Repository.find_event!(event_id)
+    event.dependencies.map { |dependency| ZyncEvent.create(event, dependency, metadata: { skip_background_sync: true }) }
+  end
+
+  def publish_dependency_events(dependency_events)
+    dependency_events.each do |dependent_event|
+      publisher.call(dependent_event, 'zync')
+    end
+  end
+
+  def enqueue_dependency_batch(event_id, dependency_events)
+    batch = Sidekiq::Batch.new
+    batch.description = "[ZyncWorker] Syncing dependencies first"
+    batch.on(:complete, self.class, { 'event_id' => event_id })
+    batch.jobs { enqueue_dependency_jobs(dependency_events) }
+  end
 
-    dependencies = event.dependencies.map do |dependency|
-      ZyncEvent.create(event, dependency)
+  def enqueue_dependency_jobs(dependency_events)
+    dependency_events.each do |dependent_event|
+      perform_async(dependent_event.event_id, dependent_event.data)
     end
+  end
 
-    dependencies.each { |dependency| publisher.call(dependency, 'zync') }
+  def on_complete(_bid, options)
+    event = EventStore::Repository.find_event!(options['event_id'])
+    perform_async(event.event_id, event.data) if event
```

Comment:
I think that `find_event!` raises an error if it doesn't exist (not sure). Which makes sense because if it reaches `on_complete` for the exact same event, we expect it to exist. But then we shouldn't need `if event` here because if it doesn't exist, it raises error and doesn't reach this line.